js/src/mesh.ts

```diff
@@ -266,23 +332,67 @@ class MeshView extends widgets.WidgetView {
         if (this.model.get("line_material")) {
             this.line_material_rgb.copy(this.model.get("line_material").obj);
         }
-        this.material_rgb.defines = {USE_RGB: true};
+
+        // update material defines in order to run correct shader code
+        this.material.defines = {USE_COLOR: true};
+        this.material_rgb.defines = {USE_RGB: true, USE_COLOR: true};
         this.line_material.defines = {AS_LINE: true};
-        this.line_material_rgb.defines = {AS_LINE: true, USE_RGB: true};
+        this.line_material_rgb.defines = {AS_LINE: true, USE_RGB: true, USE_COLOR: true};
         this.material.extensions = {derivatives: true};
+
         // locally and the visible with this object's visible trait
         this.material.visible = this.material.visible && this.model.get("visible");
         this.material_rgb.visible = this.material.visible && this.model.get("visible");
         this.line_material.visible = this.line_material.visible && this.model.get("visible");
         this.line_material_rgb.visible = this.line_material.visible && this.model.get("visible");
+
+        this.lighting_model = this.model.get("lighting_model");
         this.materials.forEach((material) => {
-            material.vertexShader = require("raw-loader!../glsl/mesh-vertex.glsl");
-            material.fragmentShader = require("raw-loader!../glsl/mesh-fragment.glsl");
             material.uniforms = this.uniforms;
+            if(this.lighting_model === this.LIGHTING_MODELS.DEFAULT) {
+                material.vertexShader = require("raw-loader!../glsl/mesh-vertex.glsl");
+                material.fragmentShader = require("raw-loader!../glsl/mesh-fragment.glsl");
+            }
+            else if(this.lighting_model === this.LIGHTING_MODELS.LAMBERT) {
+                material.vertexShader = require("raw-loader!../glsl/mesh-vertex-lambert.glsl");
+                material.fragmentShader = require("raw-loader!../glsl/mesh-fragment-lambert.glsl");
+            }
+            else if(this.lighting_model === this.LIGHTING_MODELS.PHONG) {
+                material.vertexShader = require("raw-loader!../glsl/mesh-vertex-phong.glsl");
+                material.fragmentShader = require("raw-loader!../glsl/mesh-fragment-phong.glsl");
+            }
+            else if(this.lighting_model === this.LIGHTING_MODELS.PHYSICAL) {
+                material.vertexShader = require("raw-loader!../glsl/mesh-vertex-physical.glsl");
+                material.fragmentShader = require("raw-loader!../glsl/mesh-fragment-physical.glsl");
+            }
```

Comment:
I think that instead of having 4 shader, we can put them in 1, and use the pre-processor + defines to control the specifics.

Reply:
Sounds good. Do you think that support for Lambert, Phong and Phyisical lighting models is exhaustive?

Reply:
I have recently seen dome eyed lighting. Maybe another one we can consider. It's e.g. in VTK/pyvista.

Reply:
merged into mesh-fragment and mesh-vertex